lib/Backends/NNPI/NNPIResource.cpp

```diff
@@ -544,13 +545,24 @@ bool NNPIResource::updateHostResourceFromTensor(Tensor *t, bool partialTensor) {
 
   // Pad with zeros if needed.
   if (partialData && !partialTensor) {
-    memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
-           paddedSize - unpaddedSize);
+    if (!requiresNZPadding) {
+      memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
+             paddedSize - unpaddedSize);
+    } else {
+      LOG_AND_RETURN_IF(
+          ERROR, !downcastInt64,
+          "Non-zero padding is not supported for none-index types", false);
+      int numElements = unpaddedSize / sizeof(int32_t);
+      int totalElements = paddedSize / sizeof(int32_t);
+      std::fill(reinterpret_cast<int32_t *>(hostPtr_) + numElements,
+                reinterpret_cast<int32_t *>(hostPtr_) + totalElements,
+                reinterpret_cast<const int32_t *>(hostPtr_)[numElements - 1]);
+    }
+  } else if (partialData) {
+    // Update partial size
+    partialSize_ = unpaddedSize;
```

Comment:
partialSize_ needs to be updated regardless (i.e. in this version 0 if not padding).
Later parts in the code are relying on it.
So not updating will carry over padded size from one inference to the next.